neuralop/layers/embeddings.py

```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+
+        """
+        super().__init__()
+
+        self._B = nn.Parameter(torch.randn((in_channels, mapping_size)) * scale,
+                               requires_grad=learnable)
+
+    def forward(self, x):
+        if len(x.shape) == 2:
+            x = x.unsqueeze(0)
+        batches, num_of_points, channels = x.shape
+
+        # Make shape compatible for matmul with _B.
+        # From [B, N, C] to [(B*N), C].
+        x = x.view(-1, channels)
+
+        x = x @ self._B.to(x.device)
+
+        # From [(B*N), C] to [B, N, C]
+        x = x.view(batches, num_of_points, -1)
+
+        x = 2 * torch.pi * x
+
+        return torch.cat([torch.sin(x), torch.cos(x)], dim=-1)
+
+
+class Sine(nn.Module):
+    def __init__(self, w0=1.):
+        super().__init__()
+        self.w0 = w0
+
+    def forward(self, x):
+        return torch.sin(self.w0 * x)
+
+
+class Siren(nn.Module):
+    def __init__(self,
+                 dim_in,
+                 dim_out,
+                 w0=1.,
+                 c=6.,
+                 is_first=False,
+                 use_bias=True,
+                 activation=None):
+        """
+            code modified from:
+                 https://github.com/lucidrains/siren-pytorch/blob/master/siren_pytorch/siren_pytorch.py
+            SIREN paper: https://arxiv.org/abs/2006.09661
+            The Siren layer is a linear layer followed by a sine activation function.
+
+            Parameters:
+                dim_in: int, Number of input channels.
+                dim_out: int, Number of output channels.
+                w0: float, scaling factor (denominator) used to initialize the weights, by default 6.
+                c: float, scaling factor (numerator) used to initialize the weights, by default 6.
+                is_first: bool, Whether this is the first layer of the network, by default False.
+                use_bias: bool, Whether to use bias or not, by default True.
+                activation: nn.Module, Activation function to use, by default None (which uses Sine activation).
+
+        """
+
+        super().__init__()
+        self.dim_in = dim_in
+        self.is_first = is_first
+
+        weight = torch.zeros(dim_out, dim_in)
+        bias = torch.zeros(dim_out) if use_bias else None
+        self.init_(weight, bias, c=c, w0=w0)
+
+        self.weight = nn.Parameter(weight)
+        self.bias = nn.Parameter(bias) if use_bias else None
+        self.activation = Sine(w0) if activation is None else activation
+
+    def init_(self, weight, bias, c, w0):
+        dim = self.dim_in
+
+        w_std = (1 / dim) if self.is_first else (math.sqrt(c / dim) / w0)
+        weight.uniform_(-w_std, w_std)
+
+        if bias is not None:
+            bias.uniform_(-w_std, w_std)
+
+    def forward(self, x):
+        out = torch.nn.functional.linear(x, self.weight, self.bias)
+        out = self.activation(out)
+        return out
+
+
+class SirenNet(nn.Module):
```

Comment:
@dhpitt Should we move this to `layers/mlp.py`?

Reply:
That makes sense to me